### ISSUE TITLE
Very short separate traces are labeled duplicate if starting in the same second

### ISSUE THREAD

**Issue by jes-p** [SeismicData/pyasdf#44]
I'm implementing ASDF for my lab seismology/AE experiments and have run into a timescales issue. My waveforms come as ~1 Msample blocks at 20 MHz, often multiple of these blocks occur within one second. When `add_waveforms` checks for duplicate traces, the lines quoted from `__get_waveform_ds_name` truncate the times to seconds and thus reject any traces that started in the same second as duplicates.

https://github.com/SeismicData/pyasdf/blob/c976927284efdec1264d652d47e4d0048dfe6f5c/pyasdf/asdf_data_set.py#L1299-L1300

This is a three character fix: 
`"%Y-%m-%dT%H:%M:%S"` -> `"%Y-%m-%dT%H:%M:%S.%f"`

I'm new to open source contributing and this community. Should I just clone and make this fix for myself?

**Comment by krischer** [comment]
Hi,

good point and this should definitely be changed. Instead of simply always adding the microseconds might I suggest to only add them if `start` and `end` would otherwise be different?, e.g. something like

```python
fmt_string = "%Y-%m-%dT%H:%M:%S"
s = start.strftime(fmt_string)
e = end.strftime(fmt_string)

if s == e:
    fmt_string = "%Y-%m-%dT%H:%M:%S.%f"
    s = start.strftime(fmt_string)
    e = end.strftime(fmt_string)

if s == e:
    raise ValueError("Some sensible error message")

...
````

Then the files will be identical to the current ASDF files except if really necessary.

This would additionally require to increment the version number of the ASDF format. It is currently 1.0.1 and thus should probably be incremented to 1.0.2. `pyasdf` should be ready for this but it also requires changes here: http://asdf-definition.readthedocs.io

The regular expression specifying the allowable dataset names has to be changed to optionally accomodate the microseconds. This has to be done in two places:

* http://asdf-definition.readthedocs.io/en/latest/waveforms.html#waveform-traces
* A new schema for version 1.0.2 should be added here: https://github.com/SeismicData/asdf_validate/tree/master/asdf_validate/schemas



> I'm new to open source contributing and this community. Should I just clone and make this fix for myself?

Yes - basically you clone this and other repositories, make all required changes, best add some tests and then send a PR which contains your changes. Then we'll have a look and most likely merge them. I can also take care of this in the next couple of days but it would be great if you'd be open to give it a shot.

**Comment by jes-p** [comment]
Thanks for your help! I'm taking a shot in the PR's linked by references above.

**Comment by krischer** [comment]
Thanks a lot! All looks pretty good. Before I start merging these things - I have no experience with lab data but do you think it might make sense to directly use nanoseconds instead of microseconds?

**Comment by jes-p** [comment]
Based on my lab work and what I know about others', I find it hard to imagine a need to store a waveform shorter than 1 microsecond. However, it would certainly be even more adaptable if it used nanosecond precision. Is there any character limit concern on the length of a dataset name, or any other reason _not_ to include nanoseconds?

I also don't see a `strftime` [format code](https://docs.python.org/2/library/datetime.html#strftime-and-strptime-behavior) for nanoseconds. My obspy traces do show nanoseconds so I could dig for whatever is used there.

**Comment by krischer** [comment]
Sorry for the delay. I added a few commits to your changes to write nanoseconds instead of microseconds and I also added a few tests. In the end your question if there is any reason *not* to include nanoseconds convinced me. I could not think of any reason not to include them and there might be some use cases where this is interesting.

I'll release a new version of `pyasdf` within this week which will include all the changes.

Thanks a lot for everything!
